### PR TITLE
[WIP] Initial files for flecstan (FleCSI Static Analyzer)

### LINKED ISSUE #463: Static Analysis Milestone Description
- Work with Ristra team to identify gaps in our current static analysis capability, and to improve the FleCSI interface when possible.
- Implement a static analysis tool for identifying runtime errors statically.
- Integrate static analysis tool into FleCSI project (build system, test, and invocation).

### PR DESCRIPTION
**Pull request by staleyLANL**
Fix #463

### CODE REVIEW

**Comment by junghans** [comment]
There is some formatting issue in master we have to fix first.

**Comment by ktsai7** [comment]
yes... do you have clang-format 7? if you do, then it would be the easiest if you could run make format-FleCSI in your build directory for both mpi and legion build.

**Comment by junghans** [comment]
It might not even be this PR, but the one which got merge before.

**Comment by cferenba** [comment]
The PR just before this was mine, so this may be my fault.  I don't currently have clang-format working.  If someone else can direct me on how to set it up, I'll fix my part.

**Comment by staleyLANL** [comment]
I have clang-format 7.0.0. It isn't clear to me what this issue was about. I added some completely new things to flecsi/tools, and didn't change anything at all in the rest of FleCSI. Just a brand new directory with my FleCSI Static Analysis tool.

**Comment by ktsai7** [comment]
Charles, if you have linux, then you should be able to install clang-format with the appropriate package manager.

Martin, this is just a format test failure meaning there are some spacing, tabs, or other formatting problems. If you run `make format-FleCSI` in your build directory for legion and mpi and include the changed files, this should be taken care of automatically.

File: tools/flecstan/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,36 @@
+
+# CMake
+project(flecstan LANGUAGES CXX)
+cmake_minimum_required(VERSION 3.8) # probably not necessary
+
+# C++
+set(CMAKE_CXX_COMPILER g++)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
```
Comment (junghans) [inline tools/flecstan/CMakeLists.txt]: Use `set(CMAKE_CXX_STANDARD 17)` instead.

File: tools/flecstan/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread -fno-rtti")
```
Comment (junghans) [inline tools/flecstan/CMakeLists.txt]: Use `find_package(Threads REQUIRED)` instead.

File: tools/flecstan/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pedantic -Wall -s")
```
Comment (junghans) [inline tools/flecstan/CMakeLists.txt]: Why?

File: tools/flecstan/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+# Header file location
+include_directories(src)
+
+# Source files
+file(GLOB SOURCES "src/*.cc")
```
Comment (junghans) [inline tools/flecstan/CMakeLists.txt]: Globing is bad, as new file won't get picked up with re-running CMake, please use an explicit list.

File: tools/flecstan/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+# Executable
+###set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "/usr/local/bin")
```
Comment (junghans) [inline tools/flecstan/CMakeLists.txt]: Remove.

File: tools/flecstan/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+# Executable
+###set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "/usr/local/bin") 
+add_executable(flecstan ${SOURCES})
+
+# Libraries
+target_link_libraries(
+   # executable...
+   flecstan
+   # libraries...
+   clangAnalysis clangAST clangBasic clangDriver clangEdit clangFrontend
```
Comment (junghans) [inline tools/flecstan/CMakeLists.txt]: Hard-codes libraries are a big no-no, use `find_package(LLVM)` (from `cinch`) and then `${LLVM_LIBRARIES}`.

File: tools/flecstan/compile.sh [hunk 7]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
```
Comment (junghans) [inline tools/flecstan/compile.sh]: Why do we need this script at all?

File: tools/flecstan/compile.sh [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+# ------------------------
+# Setup
+# ------------------------
+
+export COMPILE='
```
Comment (junghans) [inline tools/flecstan/compile.sh]: Why `export`?

File: tools/flecstan/readme [hunk 9]
```diff
@@ -0,0 +1,50 @@
+
```
Comment (junghans) [inline tools/flecstan/readme]: Can we turn that into markdown README, i.e. `README.md`?

**Comment by staleyLANL** [comment]
Am I supposed to be able to reply to comments? Clicking in the "Reply" boxes does nothing, and doesn't let me enter text.

**Comment by staleyLANL** [comment]
OK with set(CMAKE_CXX_STANDARD 17).

Compilation fails if I replace -pthreads with find_package(Threads REQUIRED).

@junghans "Globing is bad, as new file won't get picked up with re-running CMake, please use an explicit list."
Not sure I understand. Wouldn't the opposite be true?

At some point I tried find_package(LLVM) and similar things, but was unable to persuade cmake to entirely build the thing without errors. Only the explicit list worked, but I agree, there should be a better way. I'll try again.

True, we don't need compile.sh, but it does no harm. I've written this code as part of FleCSI, but it's entirely usable as a standalone tool. Somebody, sometime in the future, might appreciate seeing what a simple compilation looks like.

I suppose export isn't needed in that script. Force of habit from writing bash scripts.

Turning readme into a README.md was going to happen. For now, I just wanted to get the basic content submitted, not just on my computer. It's all a work-in-progress. :-)

File: tools/flecstan/CMakeLists.txt (hunk 4 above)
Comment (junghans) [reply tools/flecstan/CMakeLists.txt]: Try it out, add a new `cc`, it won't get picked up without rerunning CMake, this tends to confuse people more than being helpful. Ask  @ddaniel so a background.

File: tools/flecstan/CMakeLists.txt (hunk 6 above)
Comment (junghans) [reply tools/flecstan/CMakeLists.txt]: Some distros call the libraries `libLLVM*-7.so`.

File: tools/flecstan/CMakeLists.txt (hunk 2 above)
Comment (junghans) [reply tools/flecstan/CMakeLists.txt]: Background, not all compilers call the flag `-pthread`.

**Comment by junghans** [comment]
> Am I supposed to be able to reply to comments? Clicking in the "Reply" boxes does nothing, and doesn't let me enter text.

Work fine for me, try a different browser.

**Comment by junghans** [comment]
The error on Fedora is:
```
/builds/next-generation-codes/laristra/flecsi/tools/flecstan/src/flecstan-misc.h:6:10: fatal error: clang/Sema/Sema.h: No such file or directory
 #include "clang/Sema/Sema.h"
          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
```
which means:

- [ ] A check for required header needs to be added to `CMakeLists.txt`
- [ ] Package containing `Sema.h` needs to be added to flecsi-buildenv

**Comment by junghans** [comment]
Travis says:
```
Make Error at cinch/cmake/FindLLVM.cmake:245 (if):
  if given arguments:

    "6.0.0" "VERSION_LESS"

  Unknown arguments specified
```
so that is a problem in cinch.

**Comment by staleyLANL** [comment]
Re: the CmakeLists.txt file, I tried various ways of getting the correct Clang libraries in a similar way to how I'm doing it with LLVM, but was unable to get anything to work. Maybe somebody with more CMake knowledge that I have, can take a look. For now, in the target_link_libraries, I'm still needing to list the clang* libraries explicitly.

**Comment by ipdemes** [comment]
@staleyLANL : do you still need help with this? were you able to update cinch?

**Comment by staleyLANL** [comment]
Let's see if the CMakeLists.txt change solves the problem of Fedora finding Sema.h.
Looks like @junghans dealt with the cinch issue in #175?

**Comment by ipdemes** [comment]
@staleyLANL  could you, please, merge most recent master into your branch and resolve conflicts, if any ?

**Comment by staleyLANL** [comment]
ok, done.

**Comment by junghans** [comment]
Travis:
```
/usr/bin/ld: cannot find -lclangAST
/usr/bin/ld: cannot find -lclangBasic
```

@ktsai7, can you check the docker image for the libraries?

@staleyLANL, we still need a check for that in CMake.

**Comment by ktsai7** [comment]
sure, I guess check for clangAST and clangBasic?

**Comment by ktsai7** [comment]
By the way, in gitlab ci:
```
-- Could NOT find clang (missing: CLANG_LIBRARIES) (found version "7.0.1")
CMake Warning at cinch/cmake/FindLLVM.cmake:138 (message):
  Could not find llvm-config (LLVM >= 0).  Try manually setting LLVM_CONFIG
  to the llvm-config executable of the installation to use.
Call Stack (most recent call first):
  tools/flecstan/CMakeLists.txt:13 (find_package)


CMake Error at cinch/cmake/FindLLVM.cmake:243 (STRING):
  STRING sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command.
Call Stack (most recent call first):
  tools/flecstan/CMakeLists.txt:13 (find_package)


CMake Error at cinch/cmake/FindLLVM.cmake:249 (if):
  if given arguments:

    "VERSION_LESS" "0"

  Unknown arguments specified
Call Stack (most recent call first):
  tools/flecstan/CMakeLists.txt:13 (find_package)
```

**Comment by ktsai7** [comment]
I found `libclangAST.a` and `libclangBasic.a` at `/usr/lib/llvm-6.0/lib` in the `laristra/flecsi-third-party:ubuntu` docker container

**Comment by staleyLANL** [comment]
We definitely need the clangAST and clangBasic libraries.
Not sure how best to tell Cmake this, other than by what I already have...?

**Comment by junghans** [comment]
Ok, that makes sense now. CMake doesn't find the clang libraries as they are in a non-standard path, i.e. `/usr/lib/llvm-6.0/lib`. Looking at https://github.com/laristra/cinch/blob/master/cmake/FindClang.cmake#L56, cinch has support to find these libraries already.

@staleyLANL, you need to do something like:
```
find_package(Clang COMPONENTS AST Basic...)
```
then use `${CLANG_LIBRARIES}` instead of the hard-codes list.

**Comment by junghans** [comment]
> By the way, in gitlab ci:
> 
> ```
> -- Could NOT find clang (missing: CLANG_LIBRARIES) (found version "7.0.1")
> CMake Warning at cinch/cmake/FindLLVM.cmake:138 (message):
>   Could not find llvm-config (LLVM >= 0).  Try manually setting LLVM_CONFIG
>   to the llvm-config executable of the installation to use.
> Call Stack (most recent call first):
>   tools/flecstan/CMakeLists.txt:13 (find_package)
> 
> 
> CMake Error at cinch/cmake/FindLLVM.cmake:243 (STRING):
>   STRING sub-command REGEX, mode REPLACE needs at least 6 arguments total to
>   command.
> Call Stack (most recent call first):
>   tools/flecstan/CMakeLists.txt:13 (find_package)
> 
> 
> CMake Error at cinch/cmake/FindLLVM.cmake:249 (if):
>   if given arguments:
> 
>     "VERSION_LESS" "0"
> 
>   Unknown arguments specified
> Call Stack (most recent call first):
>   tools/flecstan/CMakeLists.txt:13 (find_package)
> ```

My guess is that `llvm-config` isn't in the path or it is name `llvm-config-6` or something.

**Comment by junghans** [comment]
Error on fedora:
```
/usr/bin/ld: CMakeFiles/flecstan.dir/src/flecstan-main.cc.o: undefined reference to symbol '_ZN4llvm11raw_ostream14flush_nonemptyEv@@LLVM_7'
/usr/bin/ld: //usr/lib64/libLLVM-7.so: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```
so
```
$ c++filt _ZN4llvm11raw_ostream14flush_nonemptyEv
llvm::raw_ostream::flush_nonempty()
```
I guess there is a llvm library missing from the link line.

File: tools/flecstan/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,47 @@
+# -*- CMake -*-
+
+project(flecstan LANGUAGES CXX)
+cmake_minimum_required(VERSION 3.8)
+
+# use cmake modules from cinch
+set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../../cinch/cmake)
+
+# C++
+set(CMAKE_CXX_COMPILER g++)
```
Comment (junghans) [inline tools/flecstan/CMakeLists.txt]: That is a bit of a no-no. How would a user tell CMake to use `clang++` or g++-9`? Please remove.

**Comment by junghans** [comment]
Now the Ubuntu builds fails with a couple of missing symbols, you need to figure out what clang libraries is missing.

**Comment by staleyLANL** [comment]
I'll remove the g++.
The clang libraries are what's causing this recurring problem. Neither ${CLANG_EXPORTED_TARGETS} nor ${CLANG_LIBRARIES} seems to bring in everything that's needed.
Just FYI, g++ compiles the code several times faster than clang++ does, on my machine, and the -s (strip symbols) brought the executable from about 500M (!) to 50M. That's why both are set, but yes, the hardcoded g++ shouldn't be there.

**Comment by junghans** [comment]
It detects and uses the right clang libraries on Fedora, so it is a ubuntu specific problem. Apparently the two distributions use different link strategies and clang versions. 

My guess is that, you need to add more clang libraries on ubuntu.

**Comment by staleyLANL** [comment]
Do you mean that CMakeLists.txt in its _current_form_ on github works on Fedora, just not Ubuntu? I actually tried adding all the clang-related libraries in my /usr/local/lib/ on the clang required components line, and it still didn't work unless I hardcoded in target_link_libraries. Although, strangely enough, I only need to hardcode clangTooling in my target_link_libraries() in order for it to work on my (ubuntu) machine.

**Comment by staleyLANL** [comment]
Just FYI, g++ compiles flecstan over 11x faster than clang++ does, in case this informs what anyone wants to do in terms of building it.

**Comment by junghans** [comment]
Yes, have a look at the log of the gitlab CI, it runs Fedora and it builds, but fails at the later format stage, while the CI on Travis runs Ubuntu and fails already at the build stage.

Our build environment is based on the latest Ubuntu docker container, which is Ubuntu 18.04.
You can get it by doing:
```
docker run -it laristra/flecsi-third-party:ubuntu /bin/bash
```
and then in the container you can do:
```
git clone --recursive -b feature/staley/flecstan https://github.com/laristra/flecsi
cd flecsi/tools/flecstan
mkdir build
cd build
cmake ..
make
```
that triggers the same error.

File: tools/flecstan/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+   src/flecstan-visitor.cc
+   src/flecstan-yaml.cc
+)
+
+# Executable
+add_executable(flecstan ${flecstan_SOURCES})
+
+# Libraries
+target_link_libraries(
+   flecstan clangTooling
```
Comment (junghans) [inline tools/flecstan/CMakeLists.txt]: Use `${CLANG_LIBRARIES}` here

**Pull request closed without merging**